linker: __data_region_start equal to __data_start

tejlmand opened the pull request:
Fixes: #38591, #38207, #37861

The commit 65a2de84a9d5c535167951bf1cf610c4f7967ea5 aligned the data
linker symbol for sections and regions.

The data region symbol start has been placed outside the sections thus
being defined as the address of the region before alignment of the first
section in the data region, usually the `datas` section.

The symbol defining the start address of the data section is after
section alignment.
In most cases the address of the data region start and datas section
start will be identical, but not always.
The data region symbol is a new linker symbol and existing code has
been depending on the old data section start symbol.
Thus, the update to the use of the data region start symbol instead of
data ram start symbol thus results in a different address when the
section is aligned to a different address.

To ensure the original behavior in all cases, the data region start
address is now moved inside the data section.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

------
~DNM to allow time to examine the ld linker script generator for possible related fixup.~

tejlmand committed:
linker: __data_region_start equal to __data_start

Fixes: #38591, #38207, #37861

The commit 65a2de84a9d5c535167951bf1cf610c4f7967ea5 aligned the data
linker symbol for sections and regions.

The data region symbol start has been placed outside the sections thus
being defined as the address of the region before alignment of the first
section in the data region, usually the `datas` section.

The symbol defining the start address of the data section is after
section alignment.
In most cases the address of the data region start and datas section
start will be identical, but not always.
The data region symbol is a new linker symbol and existing code has
been depending on the old data section start symbol.
Thus, the update to the use of the data region start symbol instead of
data ram start symbol thus results in a different address when the
section is aligned to a different address.

To ensure the original behavior in all cases, the data region start
address is now moved inside the data section.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

erwango reviewed (approved):
I confirm it fixes #38591

evgeniy-paltsev commented:
I was preparing similar PR, but you was faster :)

This also 
fixes: https://github.com/zephyrproject-rtos/zephyr/issues/38207
fixes: https://github.com/zephyrproject-rtos/zephyr/issues/37861

tejlmand committed:
cmake: linker generator: group symbol from section

The root cause of #38591 was region symbols being placed before the
section description for data region.

Some region start symbols are placed before section description, other
region start symbols are placed inside the first section in the region
and thus identical to the sections own first symbol.

To support both schemes with the linker generator, a new
`SYMBOL SECTION` argument has been added to the zephyr_linker_group()
function.

The ld_script.cmake linker script generator has been updated to support
the new argument so that generated ld linker script has identical
behavior to the templated ld linker scripts.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: linker: arm: utilize zephyr_linker_group(... SYMBOL SECTION)

The root cause of #38591 was region symbols being placed before the
section description for data region.

To support both schemes with the linker generator, a new
`SYMBOL SECTION` argument has been added to the zephyr_linker_group()
function.

This commit updates the arm/linker.cmake CMake linker file to use the
new `SYMBOL SECTION` argument for the data region group and text region
group so that those two groups now behave identical to the behavior when
using the cortex_m linker.ld template.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> I was preparing similar PR, but you was faster :)
> 
> This also
> fixes: #38207
> fixes: #37861

Thanks, added to the commit message.

tejlmand commented:
Two additional commits targeted the ld linker script generator has been added to ensure the data and text region start symbol addresses are defined identical to using the ld template files.
